cmd/pod-scaler/producer.go

```diff
@@ -0,0 +1,482 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/openhistogram/circonusllhist"
+	prometheusapi "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/test-infra/prow/interrupts"
+
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+const (
+	MetricNameCPUUsage         = `container_cpu_usage_seconds_total`
+	MetricCPUUsage             = `rate(` + MetricNameCPUUsage + containerFilter + `[3m])`
+	MetricNameMemoryWorkingSet = `container_memory_working_set_bytes`
+	MetricMemoryWorkingSet     = MetricNameMemoryWorkingSet + containerFilter
+
+	containerFilter = `{container!="POD",container!=""}`
+
+	// MaxSamplesPerRequest is the maximum number of samples that Prometheus will allow a client to ask for in
+	// one request. We also use this to approximate the maximum number of samples we should be asking any one
+	// Prometheus server for at once from many requests.
+	MaxSamplesPerRequest = 11000
+)
+
+func Produce(clients map[string]prometheusapi.API, dataCache Cache) {
+	interrupts.TickLiteral(func() {
+		for name, metric := range map[string]string{
+			MetricNameCPUUsage:         MetricCPUUsage,
+			MetricNameMemoryWorkingSet: MetricMemoryWorkingSet,
+		} {
+			name := name
+			metric := metric
+			logger := logrus.WithField("metric", name)
+			cache, err := LoadCache(dataCache, name, logger)
+			if errors.Is(err, storage.ErrObjectNotExist) {
+				ranges := map[string][]TimeRange{}
+				for cluster := range clients {
+					ranges[cluster] = []TimeRange{}
+				}
+				cache = &CachedQuery{
+					Metric:          metric,
+					RangesByCluster: ranges,
+					Data:            map[model.Fingerprint]*circonusllhist.Histogram{},
+					DataByMetaData:  map[FullMetadata][]model.Fingerprint{},
+					DataByStep:      map[StepMetadata][]model.Fingerprint{},
+				}
+			}
+			now := time.Now()
+			q := querier{
+				lock: &sync.RWMutex{},
+				data: cache,
+			}
+			wg := &sync.WaitGroup{}
+			for clusterName, client := range clients {
+				metadata := &clusterMetadata{
+					logger: logger.WithField("cluster", clusterName),
+					name:   clusterName,
+					client: client,
+					lock:   &sync.RWMutex{},
+					// there's absolutely no chance Prometheus at the current scaling will ever be able
+					// to respond to large requests it's completely capable of creating, so don't even
+					// bother asking for anything larger than 1/10th of the largest request we can
+					maxSize: MaxSamplesPerRequest / 10,
+					errors:  make(chan error),
+					// there's also no chance that Prometheus will be able to handle any real concurrent
+					// request volume, so don't even bother trying to request more samples at once than
+					// a third of the maximum samples it can technically provide in one request
```

Comment:
This was stable across my testing but I don't know that it can generalize forever - I also don't know that we can do anything more intelligent, though.